The job **`aarch64-gnu-llvm-21-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/25196236737/job/73877473092) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/73877473092) [(plain)](https://github.com/rust-lang/rust/commit/6986ad2c495e39cd4133ae07c86cb329a7a6ba81/checks/73877473092/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
Executing "/scripts/stage_2_test_set1.sh"
+ /scripts/stage_2_test_set1.sh
+ '[' 1 == 1 ']'
+ echo 'PR_CI_JOB set; skipping tidy'
+ SKIP_TIDY='--skip tidy'
+ ../x.py --stage 2 test --skip tidy --skip compiler --skip src
PR_CI_JOB set; skipping tidy
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
---
test [codegen] tests/codegen-llvm/zip.rs ... ok

failures:

---- [codegen] tests/codegen-llvm/pow_known_base.rs stdout ----
------FileCheck stdout------------------------------

------FileCheck stderr------------------------------
/checkout/tests/codegen-llvm/pow_known_base.rs:51:12: error: CHECK: expected string not found in input
 // CHECK: %[[POW:.+]] = shl nuw i32 1, %exp
           ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/pow_known_base/pow_known_base.ll:47:32: note: scanning from here
 %_71.i = icmp ult i32 %exp, 32
                               ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/pow_known_base/pow_known_base.ll:54:2: note: possible intended match here
 %1 = shl nsw i32 -1, %exp
 ^

Input file: /checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/pow_known_base/pow_known_base.ll
Check file: /checkout/tests/codegen-llvm/pow_known_base.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = 'pow_known_base.98717cbbfaeb9c42-cgu.0' 
            2: source_filename = "pow_known_base.98717cbbfaeb9c42-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32" 
            4: target triple = "aarch64-unknown-linux-gnu" 
            5:  
            6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
            7: define noundef range(i32 0, -2147483647) i32 @pow2(i32 noundef %exp) unnamed_addr #0 { 
            8: start: 
            9:  %_45.i = icmp ult i32 %exp, 32 
           10:  %0 = shl nuw i32 1, %exp 
           11:  %_9.sroa.0.0.i = select i1 %_45.i, i32 %0, i32 0 
           12:  ret i32 %_9.sroa.0.0.i 
           13: } 
           14:  
           15: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           16: define noundef range(i32 0, -2147483647) i32 @pow4(i32 noundef %exp) unnamed_addr #0 { 
           17: start: 
           18:  %_43.1.i = icmp slt i32 %exp, 0 
           19:  %_43.0.i = shl i32 %exp, 1 
           20:  %_45.i = icmp ult i32 %_43.0.i, 32 
           21:  %0 = shl nuw i32 1, %_43.0.i 
           22:  %_9.sroa.0.0.i = select i1 %_45.i, i32 %0, i32 0 
           23:  %_0.sroa.0.0.i = select i1 %_43.1.i, i32 0, i32 %_9.sroa.0.0.i, !prof !2 
           24:  ret i32 %_0.sroa.0.0.i 
           25: } 
           26:  
           27: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           28: define noundef range(i32 0, -2147483647) i32 @pow16(i32 noundef %exp) unnamed_addr #0 { 
           29: start: 
           30:  %_43.1.i = icmp ugt i32 %exp, 1073741823 
           31:  %_43.0.i = shl i32 %exp, 2 
           32:  %_45.i = icmp ult i32 %_43.0.i, 32 
           33:  %0 = shl nuw i32 1, %_43.0.i 
           34:  %_9.sroa.0.0.i = select i1 %_45.i, i32 %0, i32 0 
           35:  %_0.sroa.0.0.i = select i1 %_43.1.i, i32 0, i32 %_9.sroa.0.0.i, !prof !2 
           36:  ret i32 %_0.sroa.0.0.i 
           37: } 
           38:  
           39: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           40: define noundef range(i32 -2147483648, 1073741825) i32 @pow_minus_2(i32 noundef %exp) unnamed_addr #0 { 
           41: start: 
           42:  %_14.i = and i32 %exp, 1 
           43:  %.not.i = icmp eq i32 %_14.i, 0 
           44:  br i1 %.not.i, label %bb8.i, label %bb5.i 
           45:  
           46: bb8.i: ; preds = %start 
           47:  %_71.i = icmp ult i32 %exp, 32 
check:51'0                                    X error: no match found
           48:  %0 = shl nuw nsw i32 1, %exp 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           49:  %_19.sroa.0.0.i = select i1 %_71.i, i32 %0, i32 0 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           50:  br label %_RNvMs0_NtCsjKeDKyqFnhH_4core3numl15overflowing_pow.exit 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51:  
check:51'0     ~
           52: bb5.i: ; preds = %start 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~
           53:  %_61.i = icmp ugt i32 %exp, 31 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           54:  %1 = shl nsw i32 -1, %exp 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:51'1      ?                          possible intended match
           55:  %_15.sroa.0.0.i = select i1 %_61.i, i32 0, i32 %1 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           56:  br label %_RNvMs0_NtCsjKeDKyqFnhH_4core3numl15overflowing_pow.exit 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           57:  
check:51'0     ~
           58: _RNvMs0_NtCsjKeDKyqFnhH_4core3numl15overflowing_pow.exit: ; preds = %bb8.i, %bb5.i 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           59:  %_0.sroa.0.0.i = phi i32 [ %_15.sroa.0.0.i, %bb5.i ], [ %_19.sroa.0.0.i, %bb8.i ] 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           60:  ret i32 %_0.sroa.0.0.i 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~
           61: } 
check:51'0     ~~
           62:  
check:51'0     ~
           63: attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable "frame-pointer"="non-leaf" "probe-stack"="inline-asm" "target-cpu"="generic" "target-features"="+v8a,+outline-atomics" } 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           64:  
check:51'0     ~
           65: !llvm.module.flags = !{!0} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           66: !llvm.ident = !{!1} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~
           67:  
check:51'0     ~
           68: !0 = !{i32 8, !"PIC Level", i32 2} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           69: !1 = !{!"rustc version 1.97.0-nightly (a8080f491 2026-05-01)"} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           70: !2 = !{!"branch_weights", !"expected", i32 1, i32 2000} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>>>>>>

------------------------------------------

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-21/bin/FileCheck" "--input-file" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/pow_known_base/pow_known_base.ll" "/checkout/tests/codegen-llvm/pow_known_base.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen-llvm/pow_known_base.rs:51:12: error: CHECK: expected string not found in input
 // CHECK: %[[POW:.+]] = shl nuw i32 1, %exp
           ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/pow_known_base/pow_known_base.ll:47:32: note: scanning from here
 %_71.i = icmp ult i32 %exp, 32
                               ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/pow_known_base/pow_known_base.ll:54:2: note: possible intended match here
 %1 = shl nsw i32 -1, %exp
 ^

Input file: /checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/pow_known_base/pow_known_base.ll
Check file: /checkout/tests/codegen-llvm/pow_known_base.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = 'pow_known_base.98717cbbfaeb9c42-cgu.0' 
            2: source_filename = "pow_known_base.98717cbbfaeb9c42-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32" 
            4: target triple = "aarch64-unknown-linux-gnu" 
            5:  
            6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
            7: define noundef range(i32 0, -2147483647) i32 @pow2(i32 noundef %exp) unnamed_addr #0 { 
            8: start: 
            9:  %_45.i = icmp ult i32 %exp, 32 
           10:  %0 = shl nuw i32 1, %exp 
           11:  %_9.sroa.0.0.i = select i1 %_45.i, i32 %0, i32 0 
           12:  ret i32 %_9.sroa.0.0.i 
           13: } 
           14:  
           15: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           16: define noundef range(i32 0, -2147483647) i32 @pow4(i32 noundef %exp) unnamed_addr #0 { 
           17: start: 
           18:  %_43.1.i = icmp slt i32 %exp, 0 
           19:  %_43.0.i = shl i32 %exp, 1 
           20:  %_45.i = icmp ult i32 %_43.0.i, 32 
           21:  %0 = shl nuw i32 1, %_43.0.i 
           22:  %_9.sroa.0.0.i = select i1 %_45.i, i32 %0, i32 0 
           23:  %_0.sroa.0.0.i = select i1 %_43.1.i, i32 0, i32 %_9.sroa.0.0.i, !prof !2 
           24:  ret i32 %_0.sroa.0.0.i 
           25: } 
           26:  
           27: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           28: define noundef range(i32 0, -2147483647) i32 @pow16(i32 noundef %exp) unnamed_addr #0 { 
           29: start: 
           30:  %_43.1.i = icmp ugt i32 %exp, 1073741823 
           31:  %_43.0.i = shl i32 %exp, 2 
           32:  %_45.i = icmp ult i32 %_43.0.i, 32 
           33:  %0 = shl nuw i32 1, %_43.0.i 
           34:  %_9.sroa.0.0.i = select i1 %_45.i, i32 %0, i32 0 
           35:  %_0.sroa.0.0.i = select i1 %_43.1.i, i32 0, i32 %_9.sroa.0.0.i, !prof !2 
           36:  ret i32 %_0.sroa.0.0.i 
           37: } 
           38:  
           39: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           40: define noundef range(i32 -2147483648, 1073741825) i32 @pow_minus_2(i32 noundef %exp) unnamed_addr #0 { 
           41: start: 
           42:  %_14.i = and i32 %exp, 1 
           43:  %.not.i = icmp eq i32 %_14.i, 0 
           44:  br i1 %.not.i, label %bb8.i, label %bb5.i 
           45:  
           46: bb8.i: ; preds = %start 
           47:  %_71.i = icmp ult i32 %exp, 32 
check:51'0                                    X error: no match found
           48:  %0 = shl nuw nsw i32 1, %exp 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           49:  %_19.sroa.0.0.i = select i1 %_71.i, i32 %0, i32 0 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           50:  br label %_RNvMs0_NtCsjKeDKyqFnhH_4core3numl15overflowing_pow.exit 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51:  
check:51'0     ~
           52: bb5.i: ; preds = %start 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~
           53:  %_61.i = icmp ugt i32 %exp, 31 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           54:  %1 = shl nsw i32 -1, %exp 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:51'1      ?                          possible intended match
           55:  %_15.sroa.0.0.i = select i1 %_61.i, i32 0, i32 %1 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           56:  br label %_RNvMs0_NtCsjKeDKyqFnhH_4core3numl15overflowing_pow.exit 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           57:  
check:51'0     ~
           58: _RNvMs0_NtCsjKeDKyqFnhH_4core3numl15overflowing_pow.exit: ; preds = %bb8.i, %bb5.i 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           59:  %_0.sroa.0.0.i = phi i32 [ %_15.sroa.0.0.i, %bb5.i ], [ %_19.sroa.0.0.i, %bb8.i ] 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           60:  ret i32 %_0.sroa.0.0.i 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~
           61: } 
check:51'0     ~~
           62:  
check:51'0     ~
           63: attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable "frame-pointer"="non-leaf" "probe-stack"="inline-asm" "target-cpu"="generic" "target-features"="+v8a,+outline-atomics" } 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           64:  
check:51'0     ~
           65: !llvm.module.flags = !{!0} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           66: !llvm.ident = !{!1} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~
           67:  
check:51'0     ~
           68: !0 = !{i32 8, !"PIC Level", i32 2} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           69: !1 = !{!"rustc version 1.97.0-nightly (a8080f491 2026-05-01)"} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           70: !2 = !{!"branch_weights", !"expected", i32 1, i32 2000} 
check:51'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>>>>>>
------------------------------------------

---- [codegen] tests/codegen-llvm/pow_known_base.rs stdout end ----

failures:
    [codegen] tests/codegen-llvm/pow_known_base.rs

test result: FAILED. 874 passed; 1 failed; 184 ignored; 0 measured; 3 filtered out; finished in 18.63s
```

</details>

--- a/library/core/src/num/int_macros.rs
+++ b/library/core/src/num/int_macros.rs
@@ -1845,11 +1845,22 @@ macro_rules! int_impl {
                       without modifying the original"]
         #[inline]
         pub const fn checked_pow(self, mut exp: u32) -> Option<Self> {
+            let mut base = self;
+            let mut acc: Self = 1;
+
+            if intrinsics::is_val_statically_known(base) && base.unsigned_abs().is_power_of_two() {
+                let k = base.unsigned_abs().ilog2();
+                let shift = try_opt!(k.checked_mul(exp));
+                return if base < 0 && (exp % 2) == 1 {
+                    (-1 as Self).shl_exact(shift)
+                } else {
+                    (1 as Self).shl_exact(shift)
+                }
+            }
+
             if exp == 0 {
                 return Some(1);
             }
-            let mut base = self;
-            let mut acc: Self = 1;
 
             if intrinsics::is_val_statically_known(exp) {
                 while exp > 1 {
@@ -3096,14 +3107,23 @@ macro_rules! int_impl {
                       without modifying the original"]
         #[inline]
         pub const fn overflowing_pow(self, mut exp: u32) -> (Self, bool) {
-            if exp == 0 {
-                return (1, false);
-            }
-
             let mut base = self;
             let mut acc: Self = 1;
             let mut overflow = false;
             let mut tmp_overflow;
+
+            if intrinsics::is_val_statically_known(base) && base.unsigned_abs().is_power_of_two() {
+                let k = base.unsigned_abs().ilog2();
+                let Some(shift) = k.checked_mul(exp) else {
+                    return (0, true)
+                };
+                let base: Self = if base < 0 && (exp % 2) != 0 { -1 } else { 1 };
+                return (base.unbounded_shl(shift), base.shl_exact(shift).is_none());
+            }
+
+            if exp == 0 {
+                return (1, false);
+            }
 
             if intrinsics::is_val_statically_known(exp) {
                 while exp > 1 {

--- a/library/core/src/num/uint_macros.rs
+++ b/library/core/src/num/uint_macros.rs
@@ -2367,11 +2367,23 @@ macro_rules! uint_impl {
                       without modifying the original"]
         #[inline]
         pub const fn checked_pow(self, mut exp: u32) -> Option<Self> {
+            let mut base = self;
+            let mut acc: Self = 1;
+
+            if intrinsics::is_val_statically_known(base) && base.is_power_of_two() {
+                // change of base:
+                // if base == 2 ** k, then
+                //    (2 ** k) ** n
+                // == 2 ** (k * n)
+                // == 1 << (k * n)
+                let k = base.ilog2();
+                let shift = try_opt!(k.checked_mul(exp));
+                return (1 as Self).checked_shl(shift);
+            }
+
             if exp == 0 {
                 return Some(1);
             }
-            let mut base = self;
-            let mut acc: Self = 1;
 
             if intrinsics::is_val_statically_known(exp) {
                 while exp > 1 {
@@ -3566,13 +3578,27 @@ macro_rules! uint_impl {
                       without modifying the original"]
         #[inline]
         pub const fn overflowing_pow(self, mut exp: u32) -> (Self, bool) {
-            if exp == 0 {
-                return (1, false);
-            }
             let mut base = self;
             let mut acc: Self = 1;
             let mut overflow = false;
             let mut tmp_overflow;
+
+            if intrinsics::is_val_statically_known(base) && base.is_power_of_two() {
+                // change of base:
+                // if base == 2 ** k, then
+                //    (2 ** k) ** n
+                // == 2 ** (k * n)
+                // == 1 << (k * n)
+                let k = base.ilog2();
+                let Some(shift) = k.checked_mul(exp) else {
+                    return (0, true)
+                };
+                return ((1 as Self).unbounded_shl(shift), shift >= Self::BITS)
+            }
+
+            if exp == 0 {
+                return (1, false);
+            }
 
             if intrinsics::is_val_statically_known(exp) {
                 while exp > 1 {

--- a/tests/codegen-llvm/pow_known_base.rs
+++ b/tests/codegen-llvm/pow_known_base.rs
@@ -1,0 +1,58 @@
+//@ compile-flags: -Copt-level=3
+// Test that `pow` can use a faster implementation when `base` is a
+// known power of two
+
+#![crate_type = "lib"]
+
+// 2 ** n == 2 ** (1 * n) == 1 << (1 * n)
+// CHECK-LABEL: @pow2
+#[no_mangle]
+pub fn pow2(exp: u32) -> u32 {
+    // CHECK: %[[OVERFLOW:.+]] = icmp ult i32 %exp, 32
+    // CHECK: %[[POW:.+]] = shl nuw i32 1, %exp
+    // CHECK: %[[RET:.+]] = select i1 %[[OVERFLOW]], i32 %[[POW]], i32 0
+    // CHECK: ret i32 %[[RET]]
+    2u32.pow(exp)
+}
+
+// 4 ** n == 2 ** (2 * n) == 1 << (2 * n)
+// CHECK-LABEL: @pow4
+#[no_mangle]
+pub fn pow4(exp: u32) -> u32 {
+    // CHECK: %[[ICMP1:.+]] = icmp slt i32 %exp, 0
+    // CHECK: %[[SHIFT_AMOUNT:.+]] = shl i32 %exp, 1
+    // CHECK: %[[ICMP2:.+]] = icmp ult i32 %[[SHIFT_AMOUNT]], 32
+    // CHECK: %[[POW:.+]] = shl nuw i32 1, %[[SHIFT_AMOUNT]]
+    // CHECK: %[[SEL:.+]] = select i1 %[[ICMP2]], i32 %[[POW]], i32 0
+    // CHECK: %[[RET:.+]] = select i1 %[[ICMP1]], i32 0, i32 %[[SEL]]
+    // CHECK: ret i32 %[[RET]]
+    4u32.pow(exp)
+}
+
+// 16 ** n == 2 ** (4 * n) == 1 << (4 * n)
+// CHECK-LABEL: @pow16
+#[no_mangle]
+pub fn pow16(exp: u32) -> u32 {
+    // CHECK: %[[ICMP1:.+]] = icmp ugt i32 %exp, 1073741823
+    // CHECK: %[[SHIFT_AMOUNT:.+]] = shl i32 %exp, 2
+    // CHECK: %[[ICMP2:.+]] = icmp ult i32 %[[SHIFT_AMOUNT]], 32
+    // CHECK: %[[POW:.+]] = shl nuw i32 1, %[[SHIFT_AMOUNT]]
+    // CHECK: %[[SEL:.+]] = select i1 %[[ICMP2]], i32 %[[POW]], i32 0
+    // CHECK: %[[RET:.+]] = select i1 %[[ICMP1]], i32 0, i32 %[[SEL]]
+    // CHECK: ret i32 %[[RET]]
+    16u32.pow(exp)
+}
+
+// (-2) ** n == (-2) ** (1 * n) == 1 << (1 * n)
+// CHECK-LABEL: @pow_minus_2
+#[no_mangle]
+pub fn pow_minus_2(exp: u32) -> i32 {
+    // CHECK: %[[IS_ODD:.+]] = and i32 %exp, 1
+    // CHECK: %[[IS_EVEN:.+]] = icmp eq i32 %[[IS_ODD]], 0
+    // CHECK: %[[BASE:.+]] = select i1 %[[IS_EVEN]], i32 1, i32 -1
+    // CHECK: %[[OVERFLOW:.+]] = icmp ult i32 %exp, 32
+    // CHECK: %[[SHIFT:.+]] = shl i32 %[[BASE]], %exp
+    // CHECK: %[[RET:.+]] = select i1 %[[OVERFLOW]], i32 %[[SHIFT]], i32 0
+    // CHECK: ret i32 %[[RET]]
+    (-2i32).pow(exp)
+}

